cuML/src/common/device_buffer.hpp

```diff
@@ -0,0 +1,176 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../../../ml-prims/src/utils.h"
+
+#include "../cuML.hpp"
+
+namespace ML {
+
+/**
+ * RAII object owning a contigous typed device buffer. The passed in allocator supports asynchronus allocation and
+ * deallocation so this can be used for temporary memory 
+ * @code{.cpp}
+ * template<typename T>
+ * void foo( cumlHandle* handle, .., cudaStream_t stream )
+ * {
+ *     ...
+ *     device_buffer<T> temp( handle->getDeviceAllocator(), stream, 0 )
+ *     
+ *     temp.resize(n, stream);
+ *     kernelA<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     kernelB<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     temp.release(stream);
+ * }
+ * @endcode
+ */
+template<typename T>
+class device_buffer
+{
+public:
+    using size_type         = std::size_t;
+    using value_type        = T;
+    using iterator          = value_type*;
+    using const_iterator    = const value_type*;
+    using reference         = T&;
+    using const_reference   = const T&;
+
+    device_buffer() = delete;
+
+    device_buffer(const device_buffer& other) = delete;
+
+    device_buffer& operator=(const device_buffer& other) = delete;
+
+    device_buffer(std::shared_ptr<deviceAllocator> allocator, cudaStream_t stream, size_type n = 0)
+        : _allocator(allocator), _size(n), _capacity(n), _data(nullptr), _stream(stream)
+    {
+        if ( _size > 0 )
+        {
+            _data = static_cast<value_type*>(_allocator->allocate( _capacity*sizeof(value_type), _stream ));
+            CUDA_CHECK( cudaStreamSynchronize( _stream ) );
+        }
+    }
+
+    ~device_buffer()
+    {
+        if ( nullptr != _data ) 
+        {
+            _allocator->deallocate( _data, _capacity*sizeof(value_type), _stream );
+        }
+    }
+
+    value_type* data()
+    {
+        return _data;
+    }
+
+    const value_type* data() const
+    {
+        return _data;
+    }
+    
+    size_type size() const
+    {
+        return _size;
+    }
+    
+    void reserve( const size_type new_capacity, cudaStream_t stream )
+    {
+        _stream = stream;
+        if ( new_capacity > _capacity )
+        {
+            value_type* new_data = static_cast<value_type*>(_allocator->allocate( new_capacity*sizeof(value_type), _stream ));
+            if ( _size > 0 ) {
+                CUDA_CHECK( cudaMemcpyAsync( new_data, _data, _size*sizeof(value_type), cudaMemcpyDeviceToDevice, _stream ) );
+            }
+            if ( nullptr != _data ) {
+                _allocator->deallocate( _data, _capacity*sizeof(value_type), _stream );
+            }
+            _data = new_data;
+            _capacity = new_capacity;
+        }
+    }
+    
+    void resize(const size_type new_size, cudaStream_t stream )
+    {
+        _stream = stream;
+        if ( _capacity < new_size )
```

Comment:
knitpick: It sounds like we could replace this whole block (and the previous line) with a call to `reserve` method?

Reply:
Good spot. Just pushed a change streamlining the implementation.